### PR TITLE
feat(cost): add time-of-day peak/off-peak pricing with cached timezone resolution and tests

### PR DESCRIPTION
**Pull request by impoiler**
## Summary

Adds time-of-day (peak/off-peak) pricing support to the cost calculation engine. Providers like DeepSeek publish discounted rates during off-peak hours (e.g. 50% off outside declared peak windows). This PR evaluates a `PeakHoursSchedule` against the request's start time and scales all usage-based charges by a configured `OffPeakCostMultiplier` when the request falls outside peak windows.

## Changes

- Added `offPeakMultiplier`, `isWithinPeakWindows`, `scaleUsageCost`, `parseClockMinutes`, and `peakHoursLocation` helpers to implement time-of-day pricing evaluation.
- The multiplier is applied in `computeCostFromInput` after all usage-based costs are computed, so a single application site covers every modality (chat, embedding, etc.).
- Flat per-request fees (`CostPerRequest`) and per-search-query fees (`SearchQueriesCost`) are explicitly excluded from the discount — only usage-based charges are scaled.
- `AdditionalCost` (guardrail and MCP sidecar costs) is also excluded; those are discounted independently through their own `computeCostFromInput` calls on their own pricing rows.
- `LookupScopes` gains a `BilledAt` field representing the request's start time. Using the start time keeps pricing deterministic across streaming and non-streaming responses and matches what users see in logs. A long stream crossing a window boundary bills entirely at its start-time rate.
- `LookupScopesFromContext` populates `BilledAt` from `BifrostContextKeyRequestStartTime`. A zero value falls back to wall-clock time at evaluation rather than mispricing as peak.
- Judge calls in `computeGuardrailJudgeCost` now inherit `BilledAt` from the parent request's scopes so they price against the same instant.
- Timezone lookups are memoized in a `sync.Map` to avoid repeated filesystem hits per priced request. Failed lookups are also memoized so a bad timezone string is not retried.
- The discount fails closed: any misconfiguration (missing schedule, missing multiplier, unknown timezone, malformed window, multiplier outside `(0, 1]`) bills at the peak (higher) rate rather than silently applying a discount.
- Midnight-wrapping windows (e.g. 22:00–02:00) are handled by checking the previous day's window against an adjusted minute offset.
- Added `cost_timeofday_test.go` covering: DeepSeek's real schedule, cache-read discounting, flat-fee exclusions, all failure-closed cases, midnight-wrapping windows, non-UTC schedules, non-text modalities, zero `BilledAt` fallback, nil scopes, `parseClockMinutes` edge cases, JSON unmarshal round-trip, and an end-to-end test loading a real datasheet file through the sync pipeline.

## Type of change

- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [ ] Documentation
- [ ] Chore/CI

## Affected areas

- [x] Core (Go)
- [ ] Transports (HTTP)
- [ ] Providers/Integrations
- [ ] Plugins
- [ ] UI (React)
- [ ] Docs

## How to test

```sh
go test ./framework/modelcatalog/datasheet/... -run TestOffPeak -v
go test ./framework/modelcatalog/datasheet/... -run TestParseClockMinutes -v
go test ./framework/modelcatalog/datasheet/... -run TestEntryUnmarshal_TimeOfDayFields -v
go test ./...
```

To exercise the discount manually, configure a model pricing row with `off_peak_cost_multiplier` and a `peak_hours` schedule, then issue requests at times inside and outside the declared windows and compare the returned cost breakdowns.

## Breaking changes

- [ ] Yes
- [x] No

## Security considerations

None. The discount logic operates entirely on pricing metadata and timestamps; no user-supplied data influences the multiplier path beyond the request start time already present in context.

## Checklist

- [ ] I read `docs/contributing/README.md` and followed the guidelines
- [x] I added/updated tests where appropriate
- [ ] I updated documentation where needed
- [x] I verified builds succeed (Go and UI)
- [ ] I verified the CI pipeline passes locally if applicable

### CODE REVIEW

**Comment by impoiler** [comment]
* **#7078** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/7078?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7054** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/7054?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6579** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6579?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6578** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6578?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6577** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6577?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6576** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6576?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6575** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6575?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6575?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#6574** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6574?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `dev`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: framework/modelcatalog/datasheet/cost.go [hunk 1]
```diff
@@ -2433,3 +2450,188 @@ func passthroughUsageToCostInput(su *schemas.BifrostPassthroughUsage) costInput
...
+		start, startOK := parseClockMinutes(w.Start)
+		end, endOK := parseClockMinutes(w.End)
+		if !startOK || !endOK || len(w.Days) == 0 {
+			continue
+		}
+		wrapped := end <= start
+		if wrapped {
+			end += 24 * 60
+		}
+		valid = true
```
[the remark this answers is not in the document]
Comment (impoiler) [reply framework/modelcatalog/datasheet/cost.go]: Valid, and the most important comment on the stack. Fixed in c69867c59.

Confirmed the behaviour you described: `valid = true` fired as soon as the clock strings parsed, before any weekday was range checked. With `Days: []int{7}` the function returned `(false, true)`, `offPeakMultiplier` read that as "evaluated, not peak", and applied the discount. That inverts the whole design of this path, where base rates are the peak prices and anything unevaluable has to over-bill.

`valid = true` now sits inside the weekday loop, after the `0..6` range check, so a window becomes usable only once a real weekday is present.

Added three cases to `TestOffPeak_FailsClosed`: `Days: []int{7}`, `Days: []int{-1}`, and two windows whose every day is out of range (`{9}` and `{-3}`), which covers the case where the flag would otherwise be set by one window and read by another.

Verified both directions. Against the old code:

```
[FAIL] TestOffPeak_FailsClosed/weekday_above_range
[FAIL] TestOffPeak_FailsClosed/weekday_below_range
[FAIL] TestOffPeak_FailsClosed/every_weekday_out_of_range_across_windows
```

All three pass with the fix, and the rest of the datasheet suite stays green.

File: framework/modelcatalog/datasheet/cost_timeofday_test.go [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+	s := testStoreWithPricing(map[string]configstoreTables.TableModelPricing{
+		makeKey("deepseek-v4-flash", "deepseek", "chat"): deepSeekPricing(),
+	})
+	resp := makeChatResponse(schemas.DeepSeek, "deepseek-v4-flash", &schemas.BifrostLLMUsage{
+		PromptTokens: 1000, CompletionTokens: 1000, TotalTokens: 2000,
+	})
+
+	const peakCost = 0.00176
+	expected := peakCost / 2
+	if peak, ok := isWithinPeakWindows(deepSeekPeakHours(), time.Now()); ok && peak {
```
[the remark this answers is not in the document]
Comment (impoiler) [reply framework/modelcatalog/datasheet/cost_timeofday_test.go]: The race is real, fixed in c69867c59, though not the way you suggested.

Agreed on the diagnosis: the test read `time.Now()` to pick its expectation and `offPeakMultiplier` read `time.Now()` again to price, so a window boundary landing between the two calls would flip the result.

I did not inject a clock. This test exists specifically to prove the wall-clock fallback works when `BilledAt` is zero, so stubbing the clock would partly stub out the behaviour under test, and it would add an injection point to the production path for one test.

Instead the test now uses two schedules that give the same answer at every instant, so it never reads the clock at all:

- `alwaysPeak`: all seven days, `00:00` to `24:00`, so every instant is peak.
- `neverPeak`: two one-minute windows that cannot match at almost any time.

Each case asserts a fixed number with no clock-derived branching. The one honest gap is that `neverPeak` does cover `00:00:00` to `00:02:00` UTC, so that subtest skips inside that two-minute window rather than reporting a false failure.

`BilledAt` is still left zero in both cases, so the fallback is still what is being exercised.

File: framework/modelcatalog/datasheet/cost_timeofday_test.go [hunk 3]
```diff
@@ -0,0 +1,567 @@
...
+		{"23:59", 1439, true},
+		{"24:00", 1440, true},
+		{"24:01", 0, false},
+		{"25:00", 0, false},
+		{"01:60", 0, false},
+		{"-1:00", 0, false},
+		{"0100", 0, false},
+		{"", 0, false},
+		{"aa:bb", 0, false},
+	}
```
[the remark this answers is not in the document]
Comment (impoiler) [reply framework/modelcatalog/datasheet/cost_timeofday_test.go:465]: Not needed

**Review by akshaydeo — DISMISSED** [review]

**Comment by akshaydeo** [comment]
### Merge activity

* **Sep 11, 3:03 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/6575).
* **Sep 11, 3:06 PM UTC**: @akshaydeo merged this pull request with [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/6575).
